<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from src/wheelchair/urdf/wheelchair.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="wheelchair">
  <!-- Motor, Camera and Lidar Simulation -->
  <!--<xacro:include filename="$(find wheelchair)/urdf/wheelchair.gazebo" />-->
  <!-- Make Chassis of Bot -->
  <link name="chassis">
    <pose>0 0 0.1 0 0 0</pose>
    <inertial>
      <mass value="60"/>
      <origin rpy=" 0 0 0" xyz="0.35 0.29 0.2"/>
      <inertia ixx="0.147116667" ixy="0" ixz="0" iyy="0.334951167" iyz="0" izz="0.3978345"/>
    </inertial>
    <collision name="collision">
      <origin rpy=" 0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/chassis_motor.dae"/>
      </geometry>
    </collision>
    <visual name="chassis_visual">
      <origin rpy=" 0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/chassis_motor.dae"/>
      </geometry>
    </visual>
  </link>
  <link name="upper_chassis">
    <inertial>
      <mass value="20"/>
      <origin rpy=" 0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.147116667" ixy="0" ixz="0" iyy="0.334951167" iyz="0" izz="0.3978345"/>
    </inertial>
    <collision>
      <origin rpy="0 0 0" xyz="-0.42 -0.29 -0.4"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/upper_chassis_no_back.dae"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="-0.42 -0.29 -0.4"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/upper_chassis_no_back.dae"/>
      </geometry>
    </visual>
  </link>
  <link name="back">
    <inertial>
      <mass value="5"/>
      <origin rpy=" 0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.147116667" ixy="0" ixz="0" iyy="0.334951167" iyz="0" izz="0.3978345"/>
    </inertial>
    <collision>
      <origin rpy="0 0 0" xyz="-0.17 -0.29 -0.45"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/back.dae"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="-0.17 -0.29 -0.45"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/back.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Left Wheel -->
  <link name="left_wheel">
    <inertial>
      <mass value="15"/>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>
    <collision>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.05" radius="0.128"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="-0.523 -0.527 -0.13"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/middle_wheel_left.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Right Wheel -->
  <link name="right_wheel">
    <inertial>
      <mass value="15"/>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>
    <collision>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.05" radius="0.128"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="-0.523 -0.051 -0.13"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/middle_wheel_right.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Right Wheel Back -->
  <link name="right_wheel_back">
    <inertial>
      <mass value="3"/>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>
    <collision>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.025" radius="0.06"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="-0.06 -0.051 -0.062"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/right_wheel_back.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Left Wheel Back -->
  <link name="left_wheel_back">
    <inertial>
      <mass value="3"/>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>
    <collision>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.025" radius="0.06"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="-0.06 -0.527 -0.062"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/left_wheel_back.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Left Wheel Support -->
  <link name="left_wheel_support">
    <inertial>
      <mass value="2"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>
    <collision>
      <origin rpy="0 0 0" xyz="-0.083 -0.527 -0.125"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/left_wheel_support.dae"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="-0.083 -0.527 -0.125"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/left_wheel_support.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Right Wheel Support -->
  <link name="right_wheel_support">
    <inertial>
      <mass value="2"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>
    <collision>
      <origin rpy="0 0 0" xyz="-0.083 -0.051 -0.125"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/right_wheel_support.dae"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="-0.083 -0.051 -0.125"/>
      <geometry>
        <mesh filename="package://wheelchair/meshes/right_wheel_support.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Project center to the ground -->
  <link name="robot_footprint"/>
  <!-- ******** Define Joints ********** -->
  <joint name="robot_footprint_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="robot_footprint"/>
    <child link="chassis"/>
  </joint>
  <joint name="upper_chassis_joint" type="prismatic">
    <!--<origin xyz="${x_upper_chassis} ${y_upper_chassis} ${z_upper_chassis}" rpy="0 0 0" />-->
    <origin rpy="0 0 0" xyz="0.42 0.29 0.38"/>
    -
		
    <parent link="chassis"/>
    <child link="upper_chassis"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" lower="0.0" upper="0.045" velocity="1.0"/>
    <dynamics damping="1.0" friction="1.0"/>
  </joint>
  <joint name="back_joint" type="revolute">
    <origin rpy="0 0 0" xyz="-0.25 0.0 0.051"/>
    <parent link="upper_chassis"/>
    <child link="back"/>
    <axis xyz="0 1 0"/>
    <limit effort="10000" lower="-0.5773" upper="0.8" velocity="1.0"/>
    <dynamics damping="1.0" friction="1.0"/>
  </joint>
  <joint name="left_wheel_front_joint" type="continuous">
    <origin rpy="0 0 0" xyz="0.523 0.527 0.13"/>
    <parent link="chassis"/>
    <child link="left_wheel"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
    <limit effort="10000" velocity="1000"/>
    <dynamics damping="1.0" friction="1.0"/>
  </joint>
  <joint name="right_wheel_front_joint" type="continuous">
    <origin rpy="0 0 0" xyz="0.523 0.051 0.13"/>
    <parent link="chassis"/>
    <child link="right_wheel"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
    <limit effort="10000" velocity="1000"/>
    <dynamics damping="1.0" friction="1.0"/>
  </joint>
  <joint name="right_wheel_support_joint" type="continuous">
    <origin rpy="0 0 0" xyz="0.083 0.051 0.125"/>
    <parent link="chassis"/>
    <child link="right_wheel_support"/>
    <axis rpy="0 0 0" xyz="0 0 1"/>
    <limit effort="10000" velocity="1000"/>
    <dynamics damping="1.0" friction="1.0"/>
  </joint>
  <joint name="left_wheel_support_joint" type="continuous">
    <origin rpy="0 0 0" xyz="0.083 0.527 0.125"/>
    <parent link="chassis"/>
    <child link="left_wheel_support"/>
    <axis rpy="0 0 0" xyz="0 0 1"/>
    <limit effort="10000" velocity="1000"/>
    <dynamics damping="1.0" friction="1.0"/>
  </joint>
  <joint name="rear_left_wheel_joint" type="continuous">
    <origin rpy="0 0 0" xyz="-0.023 0 -0.063"/>
    <parent link="left_wheel_support"/>
    <child link="left_wheel_back"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
    <limit effort="10000" velocity="1000"/>
    <dynamics damping="1.0" friction="1.0"/>
  </joint>
  <joint name="rear_right_wheel_joint" type="continuous">
    <origin rpy="0 0 0" xyz="-0.023 0 -0.063"/>
    <parent link="right_wheel_support"/>
    <child link="right_wheel_back"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
    <limit effort="10000" velocity="1000"/>
    <dynamics damping="1.0" friction="1.0"/>
  </joint>
  <!-- ******** Define Transmissions ********** -->
  <transmission name="upper_chassis_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="upper_chassis_motor">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="upper_chassis_joint">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="back_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="back_motor">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="back_joint">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="left_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="left_wheel_motor">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="left_wheel_front_joint">
      <hardwareInterface>VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="right_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="right_wheel_motor">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="right_wheel_front_joint">
      <hardwareInterface>VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <!-- Add ROS plugin so we can interact with our robot -->
  <gazebo>
    <plugin filename="libgazebo_ros_control.so" name="gazebo_ros_control">
      <robotNamespace>/</robotNamespace>
    </plugin>
  </gazebo>
</robot>

